### ISSUE TITLE
extract_section_darwin failure

### ISSUE THREAD

**Issue by abakst** [travitch/whole-program-llvm#69]
Hi,

I think there is a change with respect to byte-like objects and strings/regular expressions in Python 3. In particular, the regular expression match in extraction.py:180 throws an exception because `line` is a bytes-like object.

Moreover, otool output an extra line in plain english, which I needed to skip. My version of otool is:
~~~~
llvm-otool(1): Apple Inc. version cctools-900
Apple LLVM version 9.0.0 (clang-900.0.39.2)
~~~~

I don't know python well enough to offer a merge request, but I have fixed the problem with the following patch. I can create a merge request if it looks reasonable. It's possible that this requires more version checking.

~~~~
--- extraction.py	2018-02-26 07:54:09.000000000 -0800
+++ /nix/store/qcym14s8kwlqmr8a5sv10xdp9kl0q9dm-wllvm-1.1.3/lib/python3.6/site-packages/wllvm/extraction.py	2018-02-26 07:57:33.000000000 -0800
@@ -8,6 +8,7 @@
 import tempfile
 import shutil
 import argparse
+import codecs
 
 from .popenwrapper import Popen
 
@@ -21,7 +22,7 @@
 
 from .logconfig import logConfig
 
-
+decode_hex = codecs.getdecoder("hex_codec")
 _logger = logConfig(__name__)
 
 
@@ -123,14 +124,14 @@
 
     try:
         octets = []
-        for line in lines:
-            m = otool_hexdata.match(line)
+        for line in lines[1:]:
+            m = otool_hexdata.match(line.decode())
             if not m:
                 _logger.debug('otool output:\n\t%s\nDID NOT match expectations.', line)
                 continue
             octetline = m.group(1)
             octets.extend(octetline.split())
-        retval = ''.join(octets).decode('hex').splitlines()
+        retval = decode_hex("".join(octets))[0].splitlines()
         if not retval:
             _logger.error('%s contained no %s segment', inputFile, darwinSegmentName)
     except Exception as e:
~~~~

**Comment by ianamason** [comment]
Is this a recent change in Python? What version are we talking about?  I have the same otool but
I am stuck, by choice,  in the 2.7 python world. I thought we test travis for python 3.

**Comment by abakst** [comment]
Sorry! I'm using python 3.6.4 (which was pulled in by the nix package for wllvm). It looks like you might have Travis testing python >= 3.0 but on Linux only, while this particular failure is in the OS X specific procedure.

**Comment by ianamason** [comment]
Great @abakst  I shall look into it. Thanks for the report and the proposed solution. I will get back to you with suggestions for a pull request, so your efforts can be acknowledged.

**Comment by ianamason** [comment]
So I am still looking into this. But it currently looks good. Just wish homebrew did not rename pythons
underneath me. So it will take me a while to be sure I know what I am doing.

**Comment by ianamason** [comment]
@abakst I have incorporated your changes into our working version

https://github.com/SRI-CSL/whole-program-llvm

and updated the pip package. Let me know if it works well for you. I will push it 
upstream when I do that next.

**Comment by abakst** [comment]
@ianamason, I just cloned & installed with pip. Looks good! Thanks!

**Comment by abakst** [comment]
I suppose it's likely that this breaks on older versions of OS X...

**Comment by ianamason** [comment]
I hope not, but  if homebrew is anything to go by, the motto these days is "screw the users" :-)
